Mage.Sets/src/mage/cards/t/TheTaleOfTamiyo.java

```diff
@@ -0,0 +1,207 @@
+        this.staticText = "Mill two cards. If two cards that share a card type were milled this way, draw a card and repeat this process.";
+    }
+    private TheTaleOfTamiyoEffect1(final TheTaleOfTamiyoEffect1 effect) {
+        super(effect);
+    }
+
+    @Override
+    public TheTaleOfTamiyoEffect1 copy() {
+        return new TheTaleOfTamiyoEffect1(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+
+        if (controller != null) {
+            int possibleIterations = controller.getLibrary().size() / 2;
+            int iteration = 0;
+
+            // Stop if there are not enough cards
+            if (controller.getLibrary().size() < 2) {
+                return true;
+            }
+
+            do {
+                iteration++;
+                if (iteration > possibleIterations + 20) {
+                    // Protection against infinity loops
+                    game.setDraw(source.getControllerId());
+                    return true;
+                }
+
+                // Mill 2 cards
+                List<Card> cards = new ArrayList<>(controller
+                        .millCards(2, source, game)
+                        .getCards(game));
+
+                // Stop if less than 2 cards were milled
+                if (cards.size() < 2) {
+                    break;
+                }
+
+                // Set to hold types of the first card
+                Set<CardType> firstCardTypes = new HashSet<>(cards.get(0).getCardType());
+
+                // Check if there's at least one type in common
+                boolean typesMatch = false;
+
+                // Iterate over the second card's types to see if there is any overlap with the first card's types
+                for (CardType type : cards.get(1).getCardType()) {
+                    if (firstCardTypes.contains(type)) {
+                        typesMatch = true;
+                        break;
+                    }
+                }
+
+                // If there is a match, draw a card and continue; otherwise, break
+                if (typesMatch) {
+                    controller.drawCards(1, source, game);
+                } else {
+                    break;
+                }
+
+            } while (controller.getLibrary().size() >= 2); // Continue if there are at least 2 cards left in the library
+
+            return true;
+        }
+        return false;
+    }
+
+
+
+}
+class TheTaleOfTamiyoEffect2 extends OneShotEffect {
+
+    private final FilterCard filter;
+
+    TheTaleOfTamiyoEffect2() {
+        super(Outcome.PlayForFree);
+        this.staticText = "Exile any number of target instant, sorcery, and/or Tamiyo planeswalker cards " +
+                "from your graveyard. Copy them. You may cast any number of the copies.";
+
+        // Initialize filter
+        this.filter = new FilterCard("instant, sorcery, or Tamiyo planeswalker card");
```

Comment:
make the filter `static` and put this initialization in a `static` block after it, not in the constructor.